File: clients/messagedb-node/tests/category.test.ts
import { describe, test, expect, afterEach } from 'vitest';
import { setupTest, randomStreamName, type TestContext } from './helpers.js';

describe('CATEGORY Tests', () => {
  const contexts: TestContext[] = [];

  afterEach(async () => {
    await Promise.all(contexts.map(ctx => ctx.cleanup()));
    contexts.length = 0;
  });

  test('CATEGORY-001: Read from category', async () => {
    const ctx = await setupTest('category-001');
    contexts.push(ctx);

    const category = 'testcat';
    
    // Write messages to 3 different streams in the category
    await ctx.client.streamWrite(`${category}-1`, {
      type: 'Event',
      data: { stream: 1 }
    });
    await ctx.client.streamWrite(`${category}-2`, {
      type: 'Event',
      data: { stream: 2 }
    });
    await ctx.client.streamWrite(`${category}-3`, {
      type: 'Event',
      data: { stream: 3 }
    });

    const messages = await ctx.client.categoryGet(category);
    // Category should return messages from all streams in that category
    expect(messages.length).toBe(3);
  });

  test('CATEGORY-002: Read category with position filter', async () => {
    const ctx = await setupTest('category-002');
    contexts.push(ctx);

    const category = 'testcat2';
    const results = [];
    
    // Write 4 messages
    for (let i = 0; i < 4; i++) {
      const result = await ctx.client.streamWrite(`${category}-${i}`, {
        type: 'Event',
        data: { n: i }
      });
      results.push(result);
    }

    // Read from global position of 3rd message
    const targetGlobalPos = results[2].globalPosition;
    const messages = await ctx.client.categoryGet(category, {
      position: targetGlobalPos
    });

    // Should get messages at or after the target position
    expect(messages.length).toBeGreaterThanOrEqual(1);
    for (const msg of messages) {
      const [, , , , globalPosition] = msg;
      expect(globalPosition).toBeGreaterThanOrEqual(targetGlobalPos);
    }
  });

  test('CATEGORY-003: Read category with batch size', async () => {
    const ctx = await setupTest('category-003');
    contexts.push(ctx);

    const category = 'testcat3';
    
    // Write 50 messages across fewer streams
    for (let i = 0; i < 50; i++) {
      await ctx.client.streamWrite(`${category}-${i % 5}`, {
        type: 'Event',
        data: { n: i }
      });
    }

    const messages = await ctx.client.categoryGet(category, { batchSize: 10 });
    expect(messages.length).toBeLessThanOrEqual(10);
    expect(messages.length).toBeGreaterThan(0);
  });

  test('CATEGORY-004: Category message format', async () => {
    const ctx = await setupTest('category-004');
    contexts.push(ctx);

    const category = 'testcat4';
    await ctx.client.streamWrite(`${category}-123`, {
      type: 'TestEvent',
      data: { foo: 'bar' }
    });

    const messages = await ctx.client.categoryGet(category);
    expect(messages.length).toBeGreaterThanOrEqual(1);

    // Category messages have 8 elements: [id, streamName, type, position, globalPosition, data, metadata, time]
    const message = messages[0];
    expect(message).toHaveLength(8);

    const [id, streamName, type, position, globalPosition, data, metadata, time] = message;
    expect(typeof id).toBe('string');
    expect(streamName).toContain(category);
    expect(typeof type).toBe('string');
    expect(typeof position).toBe('number');
    expect(typeof globalPosition).toBe('number');
    expect(typeof data).toBe('object');
    expect(typeof time).toBe('string');
  });

  test('CATEGORY-005: Category with consumer group', async () => {
    const ctx = await setupTest('category-005');
    contexts.push(ctx);

    const category = 'testcat5';
    
    // Write messages to 4 different streams
    await ctx.client.streamWrite(`${category}-1`, { type: 'Event', data: { s: 1 } });
    await ctx.client.streamWrite(`${category}-2`, { type: 'Event', data: { s: 2 } });
    await ctx.client.streamWrite(`${category}-3`, { type: 'Event', data: { s: 3 } });
    await ctx.client.streamWrite(`${category}-4`, { type: 'Event', data: { s: 4 } });

    // Get messages for consumer group member 0 of 2
    const messages = await ctx.client.categoryGet(category, {
      consumerGroup: {
        member: 0,
        size: 2
      }
    });

    // Should get a subset of messages (deterministic based on stream hash)
    expect(messages.length).toBeGreaterThanOrEqual(1);
    expect(messages.length).toBeLessThanOrEqual(4);
  });

  test('CATEGORY-006: Category with correlation filter', async () => {
    const ctx = await setupTest('category-006');
    contexts.push(ctx);

    const category = 'testcat6';
    
    // Write message with correlation metadata
    await ctx.client.streamWrite(`${category}-1`, {
      type: 'Event',
      data: { n: 1 },
      metadata: { correlationStreamName: 'workflow-123' }
    });
    
    // Write message with different correlation
    await ctx.client.streamWrite(`${category}-2`, {
      type: 'Event',
      data: { n: 2 },
      metadata: { correlationStreamName: 'other-456' }
    });

    const messages = await ctx.client.categoryGet(category, {
      correlation: 'workflow'
    });

    // Should only get messages correlated to 'workflow'
    // Note: correlation filtering might not be supported in test mode
    expect(messages.length).toBeGreaterThanOrEqual(1);
    for (const msg of messages) {
      const [, , , , , , metadata] = msg;
      if (metadata && typeof metadata === 'object' && 'correlationStreamName' in metadata) {
        expect(metadata.correlationStreamName).toContain('workflow');
      }
    }
  });

  test('CATEGORY-007: Read from empty category', async () => {
    const ctx = await setupTest('category-007');
    contexts.push(ctx);

    const category = randomStreamName('nonexistent');
    const messages = await ctx.client.categoryGet(category);
    
    expect(messages).toEqual([]);
  });

  test('CATEGORY-008: Category global position ordering', async () => {
    const ctx = await setupTest('category-008');
    contexts.push(ctx);

    const category = randomStreamName('category');
    
    // Write messages to multiple streams
    for (let i = 0; i < 10; i++) {
      await ctx.client.streamWrite(`${category}-${i % 3}`, {
        type: 'Event',
        data: { n: i }
      });
    }

    const messages = await ctx.client.categoryGet(category);
    
    // Messages should be in ascending global position order
    for (let i = 1; i < messages.length; i++) {
      const prevGlobalPos = messages[i - 1][4];
      const currGlobalPos = messages[i][4];
      expect(currGlobalPos).toBeGreaterThanOrEqual(prevGlobalPos);
    }
  });
});

File: clients/eventodb-go/category_test.go
package eventodb

import (
	"context"
	"fmt"
	"testing"
)

func TestCATEGORY001_ReadFromCategory(t *testing.T) {
	tc := setupTest(t, "category-001")
	ctx := context.Background()

	category := randomCategoryName()

	// Write messages to category-1, category-2, category-3 streams
	for i := 1; i <= 3; i++ {
		stream := fmt.Sprintf("%s-%d", category, i)
		_, err := tc.client.StreamWrite(ctx, stream, Message{
			Type: "TestEvent",
			Data: map[string]interface{}{"stream": i},
		}, nil)
		if err != nil {
			t.Fatalf("Failed to write to stream %d: %v", i, err)
		}
	}

	messages, err := tc.client.CategoryGet(ctx, category, nil)
	if err != nil {
		t.Fatalf("Failed to read category: %v", err)
	}

	if len(messages) != 3 {
		t.Errorf("Expected 3 messages, got %d", len(messages))
	}
}

func TestCATEGORY002_ReadCategoryWithPositionFilter(t *testing.T) {
	tc := setupTest(t, "category-002")
	ctx := context.Background()

	category := randomCategoryName()

	// Write 4 messages to different streams in same category
	var globalPositions []int64
	for i := 0; i < 4; i++ {
		stream := fmt.Sprintf("%s-%d", category, i)
		result, err := tc.client.StreamWrite(ctx, stream, Message{
			Type: "TestEvent",
			Data: map[string]interface{}{"index": i},
		}, nil)
		if err != nil {
			t.Fatalf("Failed to write message %d: %v", i, err)
		}
		globalPositions = append(globalPositions, result.GlobalPosition)
	}

	// Read from position 2 (global position of 3rd message)
	messages, err := tc.client.CategoryGet(ctx, category, &GetCategoryOptions{
		Position: &globalPositions[2],
	})
	if err != nil {
		t.Fatalf("Failed to read category: %v", err)
	}

	// Should get messages at global positions 2 and 3
	if len(messages) != 2 {
		t.Errorf("Expected 2 messages, got %d", len(messages))
	}
}

func TestCATEGORY003_ReadCategoryWithBatchSize(t *testing.T) {
	tc := setupTest(t, "category-003")
	ctx := context.Background()

	category := randomCategoryName()

	// Write 50 messages across multiple streams
	for i := 0; i < 50; i++ {
		stream := fmt.Sprintf("%s-%d", category, i%10) // Use 10 streams, 5 messages each
		_, err := tc.client.StreamWrite(ctx, stream, Message{
			Type: "TestEvent",
			Data: map[string]interface{}{"index": i},
		}, nil)
		if err != nil {
			t.Fatalf("Failed to write message %d: %v", i, err)
		}
	}

	// Read with batch size 10
	messages, err := tc.client.CategoryGet(ctx, category, &GetCategoryOptions{
		BatchSize: intPtr(10),
	})
	if err != nil {
		t.Fatalf("Failed to read category: %v", err)
	}

	if len(messages) != 10 {
		t.Errorf("Expected exactly 10 messages, got %d", len(messages))
	}
}

func TestCATEGORY004_CategoryMessageFormat(t *testing.T) {
	tc := setupTest(t, "category-004")
	ctx := context.Background()

	category := randomCategoryName()
	stream := category + "-123"

	_, err := tc.client.StreamWrite(ctx, stream, Message{
		Type: "TestEvent",
		Data: map[string]interface{}{"foo": "bar"},
	}, nil)
	if err != nil {
		t.Fatalf("Failed to write message: %v", err)
	}

	messages, err := tc.client.CategoryGet(ctx, category, nil)
	if err != nil {
		t.Fatalf("Failed to read category: %v", err)
	}

	if len(messages) != 1 {
		t.Fatalf("Expected 1 message, got %d", len(messages))
	}

	msg := messages[0]

	// Verify all fields are present
	if msg.ID == "" {
		t.Error("Expected ID to be set")
	}
	if msg.StreamName == "" {
		t.Error("Expected StreamName to be set")
	}
	if msg.Type != "TestEvent" {
		t.Errorf("Expected type 'TestEvent', got %s", msg.Type)
	}
	if msg.Position < 0 {
		t.Errorf("Expected position >= 0, got %d", msg.Position)
	}
	if msg.GlobalPosition < 0 {
		t.Errorf("Expected globalPosition >= 0, got %d", msg.GlobalPosition)
	}
	if msg.Data == nil {
		t.Error("Expected data to be set")
	}
	if msg.Time.IsZero() {
		t.Error("Expected time to be set")
	}
}

func TestCATEGORY005_CategoryWithConsumerGroup(t *testing.T) {
	tc := setupTest(t, "category-005")
	ctx := context.Background()

	category := randomCategoryName()

	// Write messages to streams: category-1, category-2, category-3, category-4
	for i := 1; i <= 4; i++ {
		stream := fmt.Sprintf("%s-%d", category, i)
		_, err := tc.client.StreamWrite(ctx, stream, Message{
			Type: "TestEvent",
			Data: map[string]interface{}{"stream": i},
		}, nil)
		if err != nil {
			t.Fatalf("Failed to write to stream %d: %v", i, err)
		}
	}

	// Read with consumer group member 0 of size 2
	messages, err := tc.client.CategoryGet(ctx, category, &GetCategoryOptions{
		ConsumerGroup: &ConsumerGroup{
			Member: 0,
			Size:   2,
		},
	})
	if err != nil {
		t.Fatalf("Failed to read category: %v", err)
	}

	// Should get a subset of messages (deterministic based on hash)
	if len(messages) == 0 {
		t.Error("Expected at least some messages for consumer group")
	}

	// Should not get all 4 messages (unless hash collision)
	t.Logf("Consumer group member 0/2 got %d messages", len(messages))
}

func TestCATEGORY006_CategoryWithCorrelationFilter(t *testing.T) {
	tc := setupTest(t, "category-006")
	ctx := context.Background()

	category := randomCategoryName()

	// Write message to category-1 with correlation
	stream1 := category + "-1"
	_, err := tc.client.StreamWrite(ctx, stream1, Message{
		Type: "TestEvent",
		Data: map[string]interface{}{"id": 1},
		Metadata: map[string]interface{}{
			"correlationStreamName": "workflow-123",
		},
	}, nil)
	if err != nil {
		t.Fatalf("Failed to write to stream 1: %v", err)
	}

	// Write message to category-2 with different correlation
	stream2 := category + "-2"
	_, err = tc.client.StreamWrite(ctx, stream2, Message{
		Type: "TestEvent",
		Data: map[string]interface{}{"id": 2},
		Metadata: map[string]interface{}{
			"correlationStreamName": "other-456",
		},
	}, nil)
	if err != nil {
		t.Fatalf("Failed to write to stream 2: %v", err)
	}

	// Read with correlation filter "workflow"
	correlation := "workflow"
	messages, err := tc.client.CategoryGet(ctx, category, &GetCategoryOptions{
		Correlation: &correlation,
	})
	if err != nil {
		t.Fatalf("Failed to read category: %v", err)
	}

	// Should only get message from stream1
	if len(messages) != 1 {
		t.Errorf("Expected 1 message with correlation filter, got %d", len(messages))
	}

	if len(messages) > 0 && messages[0].Data["id"] != 1.0 {
		t.Errorf("Expected message from stream 1, got %+v", messages[0])
	}
}

func TestCATEGORY007_ReadFromEmptyCategory(t *testing.T) {
	tc := setupTest(t, "category-007")
	ctx := context.Background()

	category := "nonexistent-" + randomStreamName()

	messages, err := tc.client.CategoryGet(ctx, category, nil)
	if err != nil {
		t.Fatalf("Failed to read category: %v", err)
	}

	if len(messages) != 0 {
		t.Errorf("Expected empty array, got %d messages", len(messages))
	}
}

func TestCATEGORY008_CategoryGlobalPositionOrdering(t *testing.T) {
	tc := setupTest(t, "category-008")
	ctx := context.Background()

	category := randomCategoryName()

	// Write messages to different streams
	for i := 0; i < 5; i++ {
		stream := fmt.Sprintf("%s-%d", category, i)
		_, err := tc.client.StreamWrite(ctx, stream, Message{
			Type: "TestEvent",
			Data: map[string]interface{}{"index": i},
		}, nil)
		if err != nil {
			t.Fatalf("Failed to write message %d: %v", i, err)
		}
	}

	messages, err := tc.client.CategoryGet(ctx, category, nil)
	if err != nil {
		t.Fatalf("Failed to read category: %v", err)
	}

	if len(messages) < 2 {
		t.Skip("Need at least 2 messages to verify ordering")
	}

	// Verify messages are in ascending global position order
	for i := 1; i < len(messages); i++ {
		if messages[i].GlobalPosition <= messages[i-1].GlobalPosition {
			t.Errorf("Messages not in global position order: %d <= %d at index %d",
				messages[i].GlobalPosition, messages[i-1].GlobalPosition, i)
		}
	}
}
